capirca/lib/fortigate.py

```diff
@@ -0,0 +1,383 @@
+# Copyright 2019 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""Fortigate generator."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import datetime
+import six
+
+from absl import logging
+from capirca.lib import aclgenerator
+from capirca.lib import nacaddr
+
+
+_ACTION_TABLE = {
+  'accept': 'accept',
+  'deny': 'deny',
+  'reject': 'deny',
+  'reject-with-tcp-rst': 'deny',  # tcp rst not supported
+}
+
+
+class Error(Exception):
+  pass
+
+
+class FilterError(Error):
+  pass
+
+
+class FortiGateValueError(Error):
+  pass
+
+
+class FortiGateFindServiceError(Error):
+  pass
+
+
+class FortiGateDuplicateTermError(Error):
+  pass
+
+
+class FortiGatePortDoesNotExist(Error):
+  pass
+
+
+class FortigatePortMap(object):
+  """Map port numbers to service names"""
+  _PORTS_TCP = {
+      179: 'BGP',
+      53: 'DNS',
+      7: 'PING',
+      79: 'FINGER',
+      21: 'FTP',
+      70: 'GOPHER',
+      443: 'HTTPS',
+      194: 'IRC',
+      2049: 'NFS',
+      119: 'NNTP',
+      110: 'POP3',
+      1723: 'PPTP',
+      25: 'SMTP',
+      22: 'SSH',
+      517: 'TALK',
+      23: 'TELNET',
+      540: 'UUCP',
+      80: 'HTTP',
+      993: 'IMAPS',
+      3389: 'RDP',
+      3306: 'MYSQL',
+      1433: 'MS-SQL',
+      1812: 'RADIUS',
+      995: 'POP3S',
+      465: 'SMTPS',
+      389: 'LDAP',
+      69: 'TFTP'
+  }
+  _PORTS_UDP = {
+      53: 'DNS',
+      7: 'PING',
+      500: 'IKE',
+      2049: 'NFS',
+      123: 'NTP',
+      520: 'RIP',
+      161: 'SNMP',
+      162: 'snmptrap',
+      514: 'SYSLOG',
+      517: 'TALK',
+      69: 'TFTP',
+      37: 'TIMESTAMP',
+      1812: 'RADIUS',
+      67: 'DHCP'
+  }
+  _PROTO_MAP = {
+      'icmp': 'ALL_ICMP',
+      'gre': 'GRE',
+      'ip': 'ALL',
+      'tcp': _PORTS_TCP,
+      'udp': _PORTS_UDP
+  }
+
+  @staticmethod
+  def GetProtocol(protocol, port=None):
+    """
+    Converts a port number to a service name.
+    :param protocol: string representing protocol (tcp, udp, etc).
+    :param port: integer representing the port number.
+    :return: the service name of provided port-protocol
+    """
+    f_proto = FortigatePortMap._PROTO_MAP.get(protocol, None)
+    if f_proto is None:
+      raise FortiGateValueError(
+        '%r protocol is not supported by Fortigate, supported protocols = %r' % (
```

Comment:
when continuing to a new line for parameters you should add two spaces to the indentation. Please fix this here as above and elsewhere in the code.

Reply:
also this line is too long, 80 characters is the max on one line.